src/ImageSharp/Processing/Processors/Transforms/AffineProcessor.cs

```diff
@@ -0,0 +1,228 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.Helpers;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Processors
+{
+    /// <summary>
+    /// Provides the base methods to perform affine transforms on an image.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal abstract class AffineProcessor<TPixel> : ResamplingWeightedProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AffineProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="sampler">The sampler to perform the resize operation.</param>
+        protected AffineProcessor(IResampler sampler)
+            : base(sampler, 1, 1, Rectangles.DefaultRectangle) // Hack to prevent Guard throwing in base, we always set the canvas
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether to expand the canvas to fit the transformed image.
+        /// </summary>
+        public bool Expand { get; set; } = true;
+
+        /// <summary>
+        /// Returns the processing matrix used for transforming the image.
+        /// </summary>
+        /// <returns>The <see cref="Matrix3x2"/></returns>
+        protected abstract Matrix3x2 CreateProcessingMatrix();
+
+        /// <summary>
+        /// Creates a new target canvas to contain the results of the matrix transform.
+        /// </summary>
+        /// <param name="sourceRectangle">The source rectangle.</param>
+        protected virtual void CreateNewCanvas(Rectangle sourceRectangle)
+        {
+            if (this.ResizeRectangle == Rectangles.DefaultRectangle)
```

Comment:
Very strange logic through the inheritance hierarchy. Aren't we fine by always calculating `ResizeRectangle` from the affine matrix without specializations?

Reply:
In rotate, yeah, i'm gonna refactor that. The potential time saved is not worth the code complexity.

Re our resampling, fixing the dimensions is easy - a missing `MathF.Ceiling` when calculating the bounds. I cannot get it to match the ImageMagick quality though and I don't know why. 😦 

Maybe I'm using the wrong weights or something.

Reply:
I think I'm gonna drop the expand option and always expand the canvas as it's the most common use-case by far. if someone wants to crop it they can do so as a separate operation. 

It's making the maths annoying and overly complicated for our optimized methods.